src/infrastructure/utility/redis_client.py

```diff
@@ -0,0 +1,87 @@
+import datetime
+import typing
+
+import aioredis
+from config.redis import RedisSettings
+
+
+class _Cache(aioredis.Redis):
+    _storage = dict()
+
+    async def get(self, key: str):
+        now = datetime.datetime.now()
+        value, expiry = self._storage.get(key, [None, None])
+        if not value:
+            return None
+        if expiry and now > expiry:
+            return None
+        return value
+
+    async def set(self, name, value, ex=None, **kwargs):
+        now = datetime.datetime.now()
+        self._storage[name] = [
+            value,
+            (now + datetime.timedelta(seconds=ex)) if ex else None,
+        ]
+        return True
```

Comment:
I would avoid inheriting from the redis class because we will be coupled on it.
It is way much better to use aggregation with singleton pattern I guess.
Here is an example:

```python
from config import settings

class _Cache:
    """Singleton Redis cache client."""

    _initialized: bool = False
    _instance: _Cache | None = None

    def __new__(cls, *args, **kwargs) -> _Cache:
        if not cls._instance:
            cls._instance = super().__new__(cls)
        return cls._instance

    def __init__(self, redis_url: str, default_ttl: int = DEFAULT_CACHE_TTL):
        if self._initialized:
            return

        self._redis_client: Redis = from_url(redis_url)
        self._ttl: int = default_ttl  # instead of expiry / ex that you defined

        self._initialized = True

    async def set(self, namespace: str, key: Any, value: dict) -> None:
        key = self._get_key(namespace, key)
        await self._redis_client.set(key, json.dumps(value, default=str).encode())

    async def get(self, namespace: str, key: Any) -> dict | None:
        key = self._get_key(namespace, key)
        raw_data: bytes | None = await self._redis_client.get(key)

        return json.loads(raw_data.decode().replace("'", '"')) if raw_data else None

...

Cache = _Cache(settings.redis.dsn)

Reply:
Actually we could discuss it with @Tom1204

Reply:
we discussed it with @Tom1204 , and decided to remove inheritance from redis class. in _Cache. I will update it soon, so that you can comment if needed